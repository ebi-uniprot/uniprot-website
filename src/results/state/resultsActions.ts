import { action } from 'typesafe-actions';
import { Dispatch, Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import fetchData from '../../utils/fetchData';
import idx, { IDXOptional } from 'idx';
import { RootState } from '../../state/state-types';
import 'regenerator-runtime/runtime';

export const REQUEST_BATCH_OF_RESULTS = 'REQUEST_BATCH_OF_RESULTS';
export const RECEIVE_BATCH_OF_RESULTS = 'RECEIVE_BATCH_OF_RESULTS';
export const UPDATE_COLUMN_SORT = 'UPDATE_COLUMN_SORT';
export const CLEAR_RESULTS = 'CLEAR_RESULTS';
export const SWITCH_VIEW_MODE = 'SWITCH_VIEW_MODE';

export const receiveBatchOfResults = (
  url: string,
  data: any,
  nextUrl: string | undefined,
  totalNumberResults: number
) =>
  action(RECEIVE_BATCH_OF_RESULTS, {
    url,
    data,
    nextUrl,
    totalNumberResults,
    receivedAt: Date.now(),
  });

export const requestBatchOfResults = (url: string) =>
  action(REQUEST_BATCH_OF_RESULTS, { url });

const getNextUrlFromResponse = (
  link: string | null | undefined
): string | undefined => {
  if (!link) {
    return;
  }
  const re = /<([0-9a-zA-Z$\-_.+!*'(),?\/:=&%]+)>; rel="next"/;
  const match = re.exec(link);
  if (match) {
    return match[1];
  }
};

export const clearResults = () => action(CLEAR_RESULTS);

type Response = {
  data: any;
  headers: {
    ['x-totalrecords']: number;
    link: string;
  };
};

export const fetchBatchOfResults = (url: string) => async (
  dispatch: Dispatch
) => {
  dispatch(requestBatchOfResults(url));
  fetchData(url).then((response: Response) => {
    const nextUrl = getNextUrlFromResponse(idx(response, _ => _.headers.link));
    dispatch(
      receiveBatchOfResults(
        url,
        response.data,
        nextUrl,
        response.headers['x-totalrecords']
      )
    );
  });
  // .catch(error => console.error(error)); // the console creates a tslint ...
  // ... error but we want to catch this in the future
};

export const shouldFetchBatchOfResults = (url: string, state: RootState) => {
  const { isFetching, isFetched } = state.results;
  return !isFetching && !isFetched[url];
};

export const fetchBatchOfResultsIfNeeded = (url: string | undefined) => (
  dispatch: ThunkDispatch<RootState, void, Action>,
  getState: () => RootState
) => {
  if (url && shouldFetchBatchOfResults(url, getState())) {
    dispatch(fetchBatchOfResults(url));
  }
};

export const switchViewMode = () => action(SWITCH_VIEW_MODE);
