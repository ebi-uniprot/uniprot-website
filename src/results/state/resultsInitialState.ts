export enum ViewMode {
  TABLE,
  CARD,
}

export type ResultsState = {
  tableColumns: string[];
  cardColumns: string[];
  facets: any[];
  nextUrl: string;
  isFetching: boolean;
  isFetched: { [url: string]: boolean };
  results: any[];
  totalNumberResults: number;
  viewMode: ViewMode;
};

const resultsInitialState = {
  tableColumns: ['accession', 'id', 'protein_name', 'gene_names', 'organism'],
  cardColumns: ['accession', 'id', 'protein_name', 'gene_names', 'organism'],
  // cardColumns: [
  //   'accession',
  //   'id',
  //   'protein_name',
  //   'gene_names',
  //   'organism',
  //   'keyword',
  //   'cc:function',
  // ],
  results: [],
  facets: [],
  isFetching: false,
  isFetched: {},
  nextUrl: '',
  totalNumberResults: 0,
  viewMode: ViewMode.CARD,
};

export default resultsInitialState;
