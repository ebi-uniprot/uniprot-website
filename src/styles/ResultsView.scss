.datalist {
  display: flex;
  height: 100%; //needed for the data list to function
  &__column {
    width: 50%;
    overflow-y: auto;
  }
}
