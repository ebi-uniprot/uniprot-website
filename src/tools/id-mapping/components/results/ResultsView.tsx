import { useState, useEffect, useRef, useMemo, FC, ReactNode } from 'react';
import { DataTableWithLoader, Loader } from 'franklin-sites';
import { useLocation } from 'react-router-dom';

import uniProtKbConverter, {
  UniProtkbAPIModel,
} from '../../../../uniprotkb/adapters/uniProtkbConverter';
import {
  UniRefAPIModel,
  UniRefLiteAPIModel,
} from '../../../../uniref/adapters/uniRefConverter';
import { UniParcAPIModel } from '../../../../uniparc/adapters/uniParcConverter';

// columns for table views
import UniProtKBColumnConfiguration from '../../../../uniprotkb/config/UniProtKBColumnConfiguration';
import UniRefColumnConfiguration from '../../../../uniref/config/UniRefColumnConfiguration';
import UniParcColumnConfiguration from '../../../../uniparc/config/UniParcColumnConfiguration';

import useDataApi from '../../../../shared/hooks/useDataApi';
import usePrefetch from '../../../../shared/hooks/usePrefetch';
import useUserPreferences from '../../../../shared/hooks/useUserPreferences';

import toolsURLs from '../../../config/urls';

import getNextURLFromHeaders from '../../../../shared/utils/getNextURLFromHeaders';
import { getParamsFromURL } from '../../../../uniprotkb/utils/resultsUtils';
// import { getEntryPathFor } from '../../../../app/config/urls';
import idMappingConverter from '../../adapters/idMappingConverter';

import { Namespace } from '../../../../shared/types/namespaces';
import { SortDirection } from '../../../../uniprotkb/types/resultsTypes';
import { SortableColumn } from '../../../../uniprotkb/types/columnTypes';
import { Column, nsToDefaultColumns } from '../../../../shared/config/columns';
import { JobTypes } from '../../../types/toolsJobTypes';
import { IDMappingNamespace } from '../../types/idMappingServerParameters';
import {
  IDMappingSearchResults,
  Mapping,
} from '../../types/idMappingSearchResults';

// import './styles/warning.scss';
// import './styles/results-view.scss';

type APIModel = UniProtkbAPIModel | UniRefLiteAPIModel | UniParcAPIModel;

const convertRow = (row: APIModel, namespace: Namespace) => {
  switch (namespace) {
    // Main namespaces
    case Namespace.uniprotkb:
      return uniProtKbConverter(row as UniProtkbAPIModel);
    case Namespace.uniref:
      return row as UniRefLiteAPIModel;
    case Namespace.uniparc:
      return row as UniParcAPIModel;
    default:
      // eslint-disable-next-line no-console
      console.warn(`Unrecognised namespace: "${namespace}"`);
      return null;
  }
};

// Note: modify original
export const getIdKeyFor = (
  namespace: IDMappingNamespace
): ((data: Mapping) => string) => {
  switch (namespace) {
    // Main namespaces
    case Namespace.uniprotkb:
      return (data) => (data.to as UniProtkbAPIModel).primaryAccession;
    case Namespace.uniref:
      return (data) => (data.to as UniRefAPIModel).id;
    case Namespace.uniparc:
      return (data) => (data.to as UniParcAPIModel).uniParcId;
    default:
      // the data is an ID
      return (data) => data.to as string;
  }
};

// TODO: create a "Column" type to cover the different column types
// and a Column renderer type with label: string and a render definition.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const ColumnConfigurations: Partial<Record<Namespace, Map<any, any>>> = {
  [Namespace.uniprotkb]: UniProtKBColumnConfiguration,
  [Namespace.uniref]: UniRefColumnConfiguration,
  [Namespace.uniparc]: UniParcColumnConfiguration,
};

type ColumnDescriptor = {
  name: string;
  label: ReactNode;
  render: (row: APIModel) => ReactNode;
  sortable?: true;
  sorted?: SortDirection;
};
const getColumnsToDisplay = (
  namespace: Namespace,
  columns: Column[] | undefined,
  sortableColumnToSortColumn: Map<Column, string>,
  sortColumn: SortableColumn,
  sortDirection: SortDirection
): ColumnDescriptor[] =>
  columns?.map((columnName) => {
    const columnConfig = ColumnConfigurations[namespace]?.get(columnName);
    if (columnConfig) {
      const columnDescriptor = {
        label: columnConfig.label,
        name: columnName,
        render: (row: APIModel) =>
          columnConfig.render(convertRow(row, namespace)),
      };
      if (sortableColumnToSortColumn.has(columnName)) {
        return {
          ...columnDescriptor,
          sortable: true,
          sorted: columnName === sortColumn ? sortDirection : undefined,
        };
      }
      return columnDescriptor;
    }
    return {
      label: columnName,
      name: columnName,
      render: () => (
        <div className="warning">{`${columnName} has no config yet`}</div>
      ),
    };
  }) || [];

type ResultsTableProps = {
  selectedEntries: string[];
  handleEntrySelection: (rowId: string) => void;
  sortableColumnToSortColumn: Map<Column, string>;
  // For ID Mapping/Peptide search
  jobID: string;
  idMappingNamespace: IDMappingNamespace;
  jobType: JobTypes;
};

const ResultsView: FC<ResultsTableProps> = ({
  selectedEntries,
  handleEntrySelection,
  sortableColumnToSortColumn,
  jobID,
  idMappingNamespace,
  jobType,
}) => {
  const userPreferences = useUserPreferences<Column[]>(
    idMappingNamespace && (`table columns for ${idMappingNamespace}` as const),
    idMappingNamespace && nsToDefaultColumns[idMappingNamespace]
  );

  const columns = userPreferences ? userPreferences[0] : [];

  const prevColumns = useRef<Column[] | undefined>(columns);
  useEffect(() => {
    // will set it *after* the current render
    prevColumns.current = columns;
  });

  //   const history = useHistory();
  const location = useLocation();

  const { search: queryParamFromUrl } = location;
  const {
    query,
    selectedFacets,
    sortColumn,
    sortDirection,
    direct,
  } = getParamsFromURL(queryParamFromUrl);

  const urls = toolsURLs(jobType);

  const initialApiUrl = urls.resultUrl(jobID, {});

  //   const initialApiUrl = getAPIQueryUrl({
  //     idMappingNamespace,
  //     query,
  //     // TODO: Do similar things for the rest of namespaces
  //     columns,
  //     selectedFacets,
  //     // Not really interested in the facets here, so try to reduce payload
  //     facets: null,
  //     sortColumn,
  //     sortDirection,
  //     jobID,
  //     jobType,
  //   });
  const [url, setUrl] = useState(initialApiUrl);
  const [metaData, setMetaData] = useState<{
    total: number;
    nextUrl?: string;
  }>(() => ({ total: 0, nextUrl: undefined }));
  usePrefetch(metaData.nextUrl);
  const [allResults, setAllResults] = useState<Mapping[]>([]);

  useEffect(() => {
    setAllResults([]);
    setMetaData({ total: 0, nextUrl: undefined });
    setUrl(initialApiUrl);
  }, [initialApiUrl]);

  const {
    data,
    loading,
    progress,
    headers,
  } = useDataApi<IDMappingSearchResults>(url);

  // NOTE: probably not needed
  // const [getIdKey, getEntryPathForEntry] = useMemo(() => {
  const [getIdKey] = useMemo(() => {
    const getIdKey = getIdKeyFor(idMappingNamespace);
    //   const getEntryPath = getEntryPathFor(idMappingNamespace);
    //   return [getIdKey, (entry: Mapping) => getEntryPath(getIdKey(entry))];
    return [getIdKey];
  }, [idMappingNamespace]);

  // redirect to entry directly when only 1 result and query marked as "direct"
  //   useEffect(() => {
  //     if (direct && metaData.total === 1 && allResults.length === 1) {
  //       const uniqueItem = allResults[0];
  //       history.replace(getEntryPathForEntry(uniqueItem));
  //     }
  //   }, [history, direct, metaData, allResults, getEntryPathForEntry]);

  useEffect(() => {
    if (!data) {
      return;
    }
    const { results } = data;
    // TODO the generic should be based on the namespace
    const transformedData = idMappingConverter(results, idMappingNamespace);
    setAllResults((allRes) => [...allRes, ...transformedData]);
    setMetaData(() => ({
      total: +(headers?.['x-totalrecords'] || 0),
      nextUrl: getNextURLFromHeaders(headers),
    }));
  }, [data, headers, idMappingNamespace]);

  if (
    // if loading the first page of results
    loading &&
    url === initialApiUrl
    // ||
    // or we just switched namespace (a bit hacky workaround to force unmount)
    // prevNamespace.current !== idMappingNamespace ||
    // or we just changed the displayed columns (hacky too...)
    // prevColumns.current !== columns
  ) {
    return <Loader progress={progress} />;
  }

  const { total, nextUrl } = metaData;

  const handleLoadMoreRows = () => nextUrl && setUrl(nextUrl);

  //   const updateColumnSort = (columnName: string) => {
  // const sortColumn = sortableColumnToSortColumn.get(columnName as Column);
  // if (!sortColumn) {
  //   return;
  // }

  // // Change sort direction
  // const updatedSortDirection =
  //   !sortDirection || sortDirection === SortDirection.descend
  //     ? SortDirection.ascend
  //     : SortDirection.descend;

  // history.push(
  //   getLocationObjForParams({
  //     pathname: SearchResultsLocations[idMappingNamespace],
  //     query,
  //     selectedFacets,
  //     sortColumn,
  //     sortDirection: updatedSortDirection,
  //   })
  // );
  //   };

  const hasMoreData = total > allResults.length;
  const loadComponent = (
    <Loader progress={progress !== 1 ? progress : undefined} />
  );

  const columnsToDisplay = idMappingNamespace
    ? getColumnsToDisplay(
        idMappingNamespace,
        columns,
        sortableColumnToSortColumn,
        sortColumn,
        sortDirection
      )
    : [
        {
          name: 'from',
          label: 'from',
          render: (row: Mapping) => row.from,
        },
        {
          name: 'to',
          label: 'to',
          render: (row: Mapping) => row.to,
        },
      ];

  return (
    <div className="results-view">
      <DataTableWithLoader
        getIdKey={getIdKey}
        columns={columnsToDisplay}
        data={allResults}
        selected={selectedEntries}
        onSelectRow={handleEntrySelection}
        // onHeaderClick={updateColumnSort}
        onLoadMoreItems={handleLoadMoreRows}
        hasMoreData={hasMoreData}
        loaderComponent={loadComponent}
      />
    </div>
  );
};

export default ResultsView;
