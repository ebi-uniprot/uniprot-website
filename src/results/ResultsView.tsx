import React from 'react';
import { DataTable, DataList } from 'franklin-sites';
import ColumnConfiguration from '../model/ColumnConfiguration';
import '../styles/alert.scss';
import {
  SelectedEntries,
  SortableColumn,
  SortDirection,
} from './types/resultsTypes';
import UniProtCard from '../view/uniprotkb/components/UniProtCard';
import { UniProtkbAPIModel } from '../model/uniprotkb/UniProtkbConverter';
import { ViewMode } from './state/resultsInitialState';

type ResultsTableProps = {
  results: any[];
  tableColumns: string[];
  selectedEntries: SelectedEntries;
  handleEntrySelection: (rowId: string) => void;
  handleHeaderClick: (column: SortableColumn) => void;
  handleLoadMoreRows: any;
  totalNumberResults: number;
  sortColumn: SortableColumn;
  sortDirection: SortDirection;
  viewMode: ViewMode;
};

const ResultsView: React.FC<ResultsTableProps> = ({
  results = [],
  totalNumberResults,
  tableColumns,
  selectedEntries,
  handleEntrySelection,
  handleLoadMoreRows,
  handleHeaderClick,
  sortColumn,
  sortDirection,
  viewMode,
}) => {
  const columns = tableColumns.map(columnName => {
    let render;
    if (columnName in ColumnConfiguration) {
      render = (row: any) => ColumnConfiguration[columnName].render(row);
    } else {
      render = () => (
        <div className="warning">{`${columnName} has no render method`}</div>
      );
    }
    const attributes = ColumnConfiguration[columnName];
    return {
      label: attributes.label,
      name: columnName,
      render,
      sortable: columnName in SortableColumn,
      sorted: columnName === sortColumn ? sortDirection : null,
    };
  });
  const hasMoreData = totalNumberResults > results.length;
  let displayComponent = null;
  switch (viewMode) {
    case ViewMode.CARD:
      displayComponent = (
        <DataList
          idKey="primaryAccession"
          data={results}
          selectable={true}
          selected={selectedEntries}
          onSelect={handleEntrySelection}
          dataRenderer={(dataItem: UniProtkbAPIModel) => (
            <UniProtCard data={dataItem} />
          )}
          onLoadMoreItems={handleLoadMoreRows}
          hasMoreData={hasMoreData}
        />
      );
      break;
    case ViewMode.TABLE:
      displayComponent = (
        <DataTable
          idKey="primaryAccession"
          columns={columns}
          data={results}
          selectable={true}
          selected={selectedEntries}
          onSelect={handleEntrySelection}
          onHeaderClick={handleHeaderClick}
          onLoadMoreItems={handleLoadMoreRows}
          hasMoreData={hasMoreData}
        />
      );
      break;
  }

  return displayComponent;
};

export default ResultsView;
