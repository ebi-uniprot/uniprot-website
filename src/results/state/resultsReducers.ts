import { ActionType } from 'typesafe-actions';
import * as resultsActions from './resultsActions';
import resultsInitialState, {
  ResultsState,
  ViewMode,
} from './resultsInitialState';

export type ResultAction = ActionType<typeof resultsActions>;
const resultsReducers = (
  state: ResultsState = resultsInitialState,
  action: ResultAction
) => {
  switch (action.type) {
    case resultsActions.REQUEST_BATCH_OF_RESULTS:
      return {
        ...state,
        isFetching: true,
      };
    case resultsActions.RECEIVE_BATCH_OF_RESULTS:
      return {
        ...state,
        results: [...state.results, ...action.payload.data.results],
        facets: action.payload.data.facets,
        lastUpdated: action.payload.receivedAt,
        totalNumberResults: action.payload.totalNumberResults,
        isFetching: false,
        isFetched: {
          ...state.isFetched,
          [action.payload.url]: true,
        },
        nextUrl: action.payload.nextUrl || '',
      };
    case resultsActions.CLEAR_RESULTS: {
      return {
        ...state,
        results: [],
        isFetching: false,
        isFetched: {},
      };
    }
    case resultsActions.SWITCH_VIEW_MODE: {
      return {
        ...state,
        viewMode:
          state.viewMode === ViewMode.CARD ? ViewMode.TABLE : ViewMode.CARD,
      };
    }
    default:
      return state;
  }
};

export default resultsReducers;
