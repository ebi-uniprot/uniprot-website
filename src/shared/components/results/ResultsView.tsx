import React, { useState, useEffect, useRef, FC } from 'react';
import { DataTable, DataList, Loader } from 'franklin-sites';
import { useHistory, useLocation } from 'react-router-dom';

import UniProtKBCard from '../../../uniprotkb/components/results/UniProtKBCard';
import UniRefCard from '../../../uniref/components/results/UniRefCard';

import uniProtKbConverter, {
  UniProtkbAPIModel,
} from '../../../uniprotkb/adapters/uniProtkbConverter';
import { UniRefLiteAPIModel } from '../../../uniref/adapters/uniRefConverter';

import { ViewMode } from '../../../uniprotkb/state/resultsInitialState';

import apiUrls, { getAPIQueryUrl } from '../../config/apiUrls';
import UniProtKBColumnConfiguration from '../../../uniprotkb/config/UniProtKBColumnConfiguration';
import UniRefColumnConfiguration, {
  UniRefColumn,
} from '../../../uniref/config/UniRefColumnConfiguration';

import useDataApi from '../../hooks/useDataApi';
import useNS from '../../hooks/useNS';

import getNextUrlFromResponse from '../../../uniprotkb/utils/queryUtils';
import {
  getParamsFromURL,
  getLocationObjForParams,
  getSortableColumnToSortColumn,
} from '../../../uniprotkb/utils/resultsUtils';

import { Namespace } from '../../types/namespaces';
import {
  SortDirection,
  ReceivedFieldData,
} from '../../../uniprotkb/types/resultsTypes';
import {
  SortableColumn,
  UniProtKBColumn,
} from '../../../uniprotkb/types/columnTypes';

import './styles/warning.scss';
import './styles/results-view.scss';
import { AllColumns } from './ResultsContainer';

type ResultsTableProps = {
  selectedEntries: string[];
  columns: AllColumns;
  viewMode: ViewMode;
  handleEntrySelection: (rowId: string) => void;
};

const convertRow = (
  row: UniProtkbAPIModel | UniRefLiteAPIModel,
  namespace: Namespace
) => {
  switch (namespace) {
    case Namespace.uniprotkb:
      return uniProtKbConverter(row as UniProtkbAPIModel);
    case Namespace.uniref:
      return row as UniRefLiteAPIModel;
    default:
      return null;
  }
};

const getIdKey = (namespace: Namespace) => {
  switch (namespace) {
    case Namespace.uniprotkb:
      return ({ primaryAccession }: { primaryAccession: string }) =>
        primaryAccession;
    case Namespace.uniref:
      return ({ id }: { id: string }) => id;
    default:
      return null;
  }
};

// TODO create a "Column" type to cover the different column types
// and a Column renderer type with label: string and a render definition.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const ColumnConfigurations: Partial<Record<Namespace, Map<any, any>>> = {
  [Namespace.uniprotkb]: UniProtKBColumnConfiguration,
  [Namespace.uniref]: UniRefColumnConfiguration,
};

const ResultsView: FC<ResultsTableProps> = ({
  selectedEntries,
  columns,
  viewMode,
  handleEntrySelection,
}) => {
  const namespace = useNS() || Namespace.uniprotkb;
  const history = useHistory();
  const location = useLocation();

  const { search: queryParamFromUrl } = location;
  const { query, selectedFacets, sortColumn, sortDirection } = getParamsFromURL(
    queryParamFromUrl
  );

  const initialApiUrl = getAPIQueryUrl({
    namespace,
    query,
    columns,
    selectedFacets,
    sortColumn,
    sortDirection,
  });
  const [url, setUrl] = useState(initialApiUrl);
  const [metaData, setMetaData] = useState<{
    total: number;
    nextUrl: string | undefined;
  }>({ total: 0, nextUrl: undefined });
  const [allResults, setAllResults] = useState<UniProtkbAPIModel[]>([]);
  const [sortableColumnToSortColumn, setSortableColumnToSortColumn] = useState<
    Map<UniProtKBColumn | UniRefColumn, string>
  >();

  const { data, headers } = useDataApi<{
    results: UniProtkbAPIModel[];
  }>(url);
  const { data: dataResultFields } = useDataApi<ReceivedFieldData>(
    apiUrls.resultsFields
  );

  const prevViewMode = useRef<ViewMode>();
  useEffect(() => {
    prevViewMode.current = viewMode;
  });

  useEffect(() => {
    setAllResults([]);
    setMetaData({ total: 0, nextUrl: undefined });
    setUrl(initialApiUrl);
  }, [initialApiUrl]);

  useEffect(() => {
    if (!data) {
      return;
    }
    const { results } = data;
    setAllResults((allRes) => [...allRes, ...results]);
    setMetaData(() => ({
      total: headers['x-totalrecords'],
      nextUrl: getNextUrlFromResponse(headers.link),
    }));
  }, [data, headers]);

  useEffect(() => {
    if (!dataResultFields) return;
    setSortableColumnToSortColumn(
      getSortableColumnToSortColumn(dataResultFields)
    );
  }, [dataResultFields]);

  if (
    allResults.length === 0 ||
    !sortableColumnToSortColumn ||
    sortableColumnToSortColumn.size === 0 ||
    prevViewMode.current !== viewMode
  ) {
    return <Loader />;
  }

  const { total, nextUrl } = metaData;

  const handleLoadMoreRows = () => nextUrl && setUrl(nextUrl);

  const updateColumnSort = (column: SortableColumn) => {
    const sortableColumn = sortableColumnToSortColumn.get(column);
    if (!sortableColumn) return;

    // Change sort direction
    const updatedSortDirection =
      !sortDirection || sortDirection === SortDirection.descend
        ? SortDirection.ascend
        : SortDirection.descend;

    history.push(
      getLocationObjForParams({
        pathname: `/${namespace}`,
        query,
        selectedFacets,
        sortColumn: sortableColumn,
        sortDirection: updatedSortDirection,
      })
    );
  };

  const hasMoreData = total > allResults.length;
  let dataView;
  if (viewMode === ViewMode.CARD) {
    dataView = (
      <DataList
        getIdKey={getIdKey(namespace)}
        data={allResults}
        dataRenderer={(dataItem: UniProtkbAPIModel | UniRefLiteAPIModel) => {
          switch (namespace) {
            case Namespace.uniref: {
              const data = dataItem as UniRefLiteAPIModel;
              return (
                <UniRefCard
                  data={data}
                  selected={selectedEntries.includes(data.id)}
                  handleEntrySelection={handleEntrySelection}
                />
              );
            }
            case Namespace.uniprotkb:
            default: {
              const data = dataItem as UniProtkbAPIModel;
              return (
                <UniProtKBCard
                  data={data}
                  selected={selectedEntries.includes(data.primaryAccession)}
                  handleEntrySelection={handleEntrySelection}
                />
              );
            }
          }
        }}
        onLoadMoreItems={handleLoadMoreRows}
        hasMoreData={hasMoreData}
        loaderComponent={<Loader />}
        scrollDataAttribute="sidebar-content"
      />
    );
  } else {
    // viewMode === ViewMode.TABLE
    const columnsToDisplay =
      columns.map((columnName) => {
        const columnConfig = ColumnConfigurations[namespace]?.get(columnName);
        if (columnConfig) {
          return {
            label: columnConfig.label,
            name: columnName,
            render: (row: UniProtkbAPIModel | UniRefLiteAPIModel) => {
              const convertedRow = convertRow(row, namespace);
              return columnConfig.render(convertedRow);
            },
            sortable: sortableColumnToSortColumn.has(columnName),
            sorted: columnName === sortColumn && sortDirection, // TODO this doesn't seem to update the view
          };
        }
        return {
          label: columnName,
          name: columnName,
          render: () => (
            <div className="warning">{`${columnName} has no config`}</div>
          ),
          sortable: false,
          sorted: false,
        };
      }) || [];
    dataView = (
      <DataTable
        getIdKey={getIdKey(namespace)}
        columns={columnsToDisplay}
        data={allResults}
        selectable
        selected={selectedEntries}
        onSelect={handleEntrySelection}
        onHeaderClick={updateColumnSort}
        onLoadMoreItems={handleLoadMoreRows}
        hasMoreData={hasMoreData}
        loaderComponent={<Loader />}
        scrollDataAttribute="sidebar-content"
      />
    );
  }
  return <div className="results-view">{dataView}</div>;
};

export default ResultsView;
