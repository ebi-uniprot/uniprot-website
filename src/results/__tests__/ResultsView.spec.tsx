import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import ResultsView from '../ResultsView';
import { ViewMode } from '../state/resultsInitialState';

configure({ adapter: new Adapter() });

describe('ResultsView component', () => {
  test('should render table', () => {
    const props = {
      viewMode: ViewMode.TABLE,
      tableColumns: ['accession'],
      selectedEntries: {},
      results: [{ accession: '1234' }, { accession: '5678' }],
      sort: { column: 'accession', direction: 'descend' },
      handleEntrySelection: jest.fn(),
      handleHeaderClick: jest.fn(),
      handleLoadMoreRows: jest.fn(),
      totalNumberResults: 2,
      sortColumn: null,
      sortDirection: null,
    };
    const wrapper = shallow(<ResultsView {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should render cards', () => {
    const props = {
      viewMode: ViewMode.CARD,
      tableColumns: ['accession'],
      selectedEntries: {},
      results: [{ accession: '1234' }, { accession: '5678' }],
      sort: { column: 'accession', direction: 'descend' },
      handleEntrySelection: jest.fn(),
      handleHeaderClick: jest.fn(),
      handleLoadMoreRows: jest.fn(),
      totalNumberResults: 2,
      sortColumn: null,
      sortDirection: null,
    };
    const wrapper = shallow(<ResultsView {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
