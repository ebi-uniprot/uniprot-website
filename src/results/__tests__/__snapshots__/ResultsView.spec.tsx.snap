// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsView component should render cards 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ResultsView
    handleEntrySelection={[MockFunction]}
    handleHeaderClick={[MockFunction]}
    handleLoadMoreRows={[MockFunction]}
    results={
      Array [
        Object {
          "accession": "1234",
        },
        Object {
          "accession": "5678",
        },
      ]
    }
    selectedEntries={Object {}}
    sort={
      Object {
        "column": "accession",
        "direction": "descend",
      }
    }
    sortColumn={null}
    sortDirection={null}
    tableColumns={
      Array [
        "accession",
      ]
    }
    totalNumberResults={2}
    viewMode={1}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "data": Array [
        Object {
          "accession": "1234",
        },
        Object {
          "accession": "5678",
        },
      ],
      "dataRenderer": [Function],
      "hasMoreData": false,
      "idKey": "primaryAccession",
      "onLoadMoreItems": [MockFunction],
      "onSelect": [MockFunction],
      "selectable": true,
      "selected": Object {},
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "data": Array [
          Object {
            "accession": "1234",
          },
          Object {
            "accession": "5678",
          },
        ],
        "dataRenderer": [Function],
        "hasMoreData": false,
        "idKey": "primaryAccession",
        "onLoadMoreItems": [MockFunction],
        "onSelect": [MockFunction],
        "selectable": true,
        "selected": Object {},
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;

exports[`ResultsView component should render table 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ResultsView
    handleEntrySelection={[MockFunction]}
    handleHeaderClick={[MockFunction]}
    handleLoadMoreRows={[MockFunction]}
    results={
      Array [
        Object {
          "accession": "1234",
        },
        Object {
          "accession": "5678",
        },
      ]
    }
    selectedEntries={Object {}}
    sort={
      Object {
        "column": "accession",
        "direction": "descend",
      }
    }
    sortColumn={null}
    sortDirection={null}
    tableColumns={
      Array [
        "accession",
      ]
    }
    totalNumberResults={2}
    viewMode={0}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "columns": Array [
        Object {
          "label": "Entry",
          "name": "accession",
          "render": [Function],
          "sortable": true,
          "sorted": null,
        },
      ],
      "data": Array [
        Object {
          "accession": "1234",
        },
        Object {
          "accession": "5678",
        },
      ],
      "hasMoreData": false,
      "idKey": "primaryAccession",
      "onHeaderClick": [MockFunction],
      "onLoadMoreItems": [MockFunction],
      "onSelect": [MockFunction],
      "selectable": true,
      "selected": Object {},
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "columns": Array [
          Object {
            "label": "Entry",
            "name": "accession",
            "render": [Function],
            "sortable": true,
            "sorted": null,
          },
        ],
        "data": Array [
          Object {
            "accession": "1234",
          },
          Object {
            "accession": "5678",
          },
        ],
        "hasMoreData": false,
        "idKey": "primaryAccession",
        "onHeaderClick": [MockFunction],
        "onLoadMoreItems": [MockFunction],
        "onSelect": [MockFunction],
        "selectable": true,
        "selected": Object {},
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
